webapp-django/crashstats/home/static/home/js/home.js

```diff
@@ -0,0 +1,462 @@
+        return date.toISOString().substring(0, 10);
+    }
+
+    /**
+     * History change handler. Reloads the graph's data.
+     */
+    window.onpopstate = function (e) {
+        duration = e.state.duration;
+
+        $('#duration .selected').removeClass('selected');
+        $('#duration .days-' + duration).addClass('selected');
+
+        updateDates();
+        loadData();
+    };
+
+    /**
+     * Click event handler for the duration menu. Changes the history and
+     * reloads the graph's data.
+     */
+    $('#duration').on('click', 'a', function (e) {
+        e.preventDefault();
+
+        var link = $(this);
+
+        // Do nothing if this is already selected.
+        if (link.hasClass('selected')) {
+            return;
+        }
+
+        duration = link.data('duration');
+        var url = link.attr('href');
+
+        $('#duration .selected').removeClass('selected');
+        link.addClass('selected');
+
+        window.history.pushState({ duration: duration }, pageTitle, url);
+
+        updateDates();
+        loadData();
+    });
+
+    /**
+     * Change the dates of the request depending on the duration.
+     */
+    function updateDates() {
+        endDate = new Date();
+        endDate.setDate(endDate.getDate() - 1);  // We start with yesterday.
+
+        startDate = new Date();
+        startDate.setDate(endDate.getDate() - duration);
+    }
+
+    /**
+     * Get count of crashes per day and per version.
+     */
+    function querySuperSearch() {
+        var deferred = $.Deferred();
+
+        // Super Search uses datetimes instead of dates, and thus cannot
+        // use yesterday's date as the upper bound. It needs to use today's
+        // date so that results from yesterday are returned.
+        var searchEndDate = new Date();
+
+        var params = {
+            product: product,
+            version: versions,
+            date: [
+                '>=' + dateIsoFormat(startDate),
+                '<' + dateIsoFormat(searchEndDate),
+            ],
+            '_histogram.date': 'version',
+            _results_number: 0,
+        };
+
+        var url = superSearchApiUrl + '?' + $.param(params, true);
+        $.ajax({url: url})
+        .then(onSuperSearchData)
+        .done(function (data) {
+            return deferred.resolve(data);
+        })
+        .fail(onQueryError)
+        .fail(function () {
+            deferred.reject();
+        });
+
+        return deferred;
+    }
+
+    /**
+     * Get count of average daily installs per day and per version.
+     */
+    function queryADI() {
+        var deferred = $.Deferred();
+
+        var params = {
+            product: product,
+            versions: versions,
+            platforms: platforms,
+            start_date: dateIsoFormat(startDate),
+            end_date: dateIsoFormat(endDate),
+        };
+
+        var url = adiApiUrl + '?' + $.param(params, true);
+        $.ajax({url: url})
+        .then(onADIData)
```

Comment:
uh?! I didn't even know you can do that. I thought `$.ajax` returns a promise that only has `done` and `fail`.